roachtest: kv/contention/nodes=4 failed

SHA: https://github.com/cockroachdb/cockroach/commits/92fef12128c997233d985d1c19e11faac005073f

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=kv/contention/nodes=4 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1413388&tab=buildLog

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20190731-1413388/kv/contention/nodes=4/run_1
	cluster.go:1726,kv.go:197,cluster.go:2069,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1564553448-32-n5cpu4:5 -- ./workload run kv --init --secondary-index --duration=1h0m0s --cycle-length=512 --concurrency=128 --batch=4 --splits=4 {pgurl:1-4} returned:
		stderr:
		
		stdout:
		 write
		     58s        0          802.6          306.2     13.6   1342.2   9663.7  20401.1 write
		     59s        0          464.1          308.9      8.9    570.4   1342.2  12884.9 write
		    1m0s        0          164.0          306.5      6.3   1476.4   1811.9  36507.2 write
		_elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
		    1m1s        0           76.0          302.7      7.9   3087.0  15032.4  34359.7 write
		    1m2s        0          172.0          300.6      7.3   2550.1   3489.7  45097.2 write
		    1m3s        0          383.9          301.9      7.9   3087.0   4563.4  60129.5 write
		    1m4s        0          208.1          300.5      9.4   3892.3   6442.5  18253.6 write
		    1m5s        0          535.8          304.1      7.9   2013.3   3489.7   6174.0 write
		    1m6s        0          152.0          301.8      6.6   2281.7   6174.0   7516.2 write
		    1m7s        0           68.0          298.3      7.3   2415.9   5368.7  20401.1 write
		Error: EOF
		Error:  exit status 1
		: exit status 1
	cluster.go:2090,kv.go:207,test_runner.go:691: unexpected node event: 1: dead

```

```
E190731 16:28:51.015171 68369 util/log/crash_reporting.go:207  [n1,s1,r22/3:/Table/53/1/-{301330â€¦-291225â€¦}] a panic has occurred!
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xd9d4e2]

goroutine 68369 [running]:
panic(0x38abde0, 0x678abb0)
	/usr/local/go/src/runtime/panic.go:565 +0x2c5 fp=0xc009996f78 sp=0xc009996ee8 pc=0x787af5
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc00067a000, 0x4583800, 0xc00b374d50)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:181 +0x121 fp=0xc009996fd8 sp=0xc009996f78 pc=0x12a3651
runtime.call32(0x0, 0x3e5b948, 0xc0073b0c60, 0x1800000018)
	/usr/local/go/src/runtime/asm_amd64.s:519 +0x3b fp=0xc009997008 sp=0xc009996fd8 pc=0x7b5b7b
panic(0x38abde0, 0x678abb0)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5 fp=0xc009997098 sp=0xc009997008 pc=0x7879e5
runtime.panicmem(...)
	/usr/local/go/src/runtime/panic.go:82
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:390 +0x411 fp=0xc0099970c8 sp=0xc009997098 pc=0x79d651
github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace.(*traceSet).Remove(0x0, 0xc009571d40)
	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace/trace.go:478 +0x22 fp=0xc0099970f0 sp=0xc0099970c8 pc=0xd9d4e2
github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace.(*trace).Finish(0xc009571d40)
	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace/trace.go:412 +0x137 fp=0xc0099971c0 sp=0xc0099970f0 pc=0xd9d097
github.com/cockroachdb/cockroach/pkg/util/tracing.(*span).FinishWithOptions(0xc00999a0e0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/tracer_span.go:316 +0x131 fp=0xc009997228 sp=0xc0099971c0 pc=0xecdf91
github.com/cockroachdb/cockroach/pkg/util/tracing.(*span).Finish(0xc00999a0e0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/tracer_span.go:300 +0x67 fp=0xc009997288 sp=0xc009997228 pc=0xecde47
github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).Send(0xc000564360, 0x4583800, 0xc00b374d50, 0x15b68acc61193135, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:773 +0xc3b fp=0xc009997768 sp=0xc009997288 pc=0x1bd9d3b
github.com/cockroachdb/cockroach/pkg/internal/client.(*CrossRangeTxnWrapperSender).Send(0xc00068a160, 0x4583800, 0xc00b374d50, 0x15b68acc61193135, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:217 +0xaf fp=0xc009997850 sp=0xc009997768 pc=0x12e3c3f
github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).sendUsingSender(0xc00068a100, 0x4583800, 0xc00b374d50, 0x15b68acc61193135, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:665 +0x122 fp=0xc009997938 sp=0xc009997850 pc=0x12e6462
```

The timing of this points towards it having the exact same cause as https://github.com/cockroachdb/cockroach/issues/39188.

--- a/pkg/storage/cmd_app_batch.go
+++ b/pkg/storage/cmd_app_batch.go
@@ -93,7 +93,7 @@ func (b *cmdAppBatch) add(e *raftpb.Entry, d decodedRaftEntry) {
 // numEmptyEntries indicates the number of entries in the consumed portion of
 // toProcess contained a zero-byte payload.
 func (b *cmdAppBatch) decode(
-	ctx context.Context, toProcess []raftpb.Entry, decodeBuf *decodedRaftEntry,
+	ctx context.Context, toProcess []raftpb.Entry, decodeBuf *decodedRaftEntry, maxEntries int,
 ) (
 	foundNonTrivialEntry bool,
 	numEmptyEntries int,
@@ -101,7 +101,7 @@ func (b *cmdAppBatch) decode(
 	errExpl string,
 	err error,
 ) {
-	for len(toProcess) > 0 {
+	for len(toProcess) > 0 && (maxEntries <= 0 || int(b.cmdBuf.len) < maxEntries) {
 		e := &toProcess[0]
 		if len(e.Data) == 0 {
 			numEmptyEntries++

--- a/pkg/storage/replica_application.go
+++ b/pkg/storage/replica_application.go
@@ -164,6 +164,16 @@ func (r *Replica) retrieveLocalProposals(ctx context.Context, b *cmdAppBatch) {
 	for ok := it.init(&b.cmdBuf); ok; ok = it.next() {
 		cmd := it.cur()
 		cmd.proposal = r.mu.proposals[cmd.idKey]
+		if cmd.proposedLocally() && cmd.raftCmd.MaxLeaseIndex != cmd.proposal.command.MaxLeaseIndex {
+			// If this entry does not have the most up-to-date view of the
+			// corresponding proposal's maximum lease index then the proposal
+			// must have been reproposed with a higher lease index. (see
+			// tryReproposeWithNewLeaseIndex). In that case, there's a newer
+			// version of the proposal in the pipeline, so don't consider this
+			// entry to have been proposed locally. The entry must necessarily be
+			// rejected by checkForcedErr.
+			cmd.proposal = nil
+		}
 		if cmd.proposedLocally() {
 			// We initiated this command, so use the caller-supplied context.
 			cmd.ctx = cmd.proposal.ctx
@@ -179,16 +189,7 @@ func (r *Replica) retrieveLocalProposals(ctx context.Context, b *cmdAppBatch) {
 	for ok := it.init(&b.cmdBuf); ok; ok = it.next() {
 		cmd := it.cur()
 		toRelease := int64(0)
-		shouldRemove := cmd.proposedLocally() &&
-			// If this entry does not have the most up-to-date view of the
-			// corresponding proposal's maximum lease index then the proposal
-			// must have been reproposed with a higher lease index. (see
-			// tryReproposeWithNewLeaseIndex). In that case, there's a newer
-			// version of the proposal in the pipeline, so don't remove the
-			// proposal from the map. We expect this entry to be rejected by
-			// checkForcedErr.
-			cmd.raftCmd.MaxLeaseIndex == cmd.proposal.command.MaxLeaseIndex
-		if shouldRemove {
+		if cmd.proposedLocally() {
 			// Delete the proposal from the proposals map. There may be reproposals
 			// of the proposal in the pipeline, but those will all have the same max
 			// lease index, meaning that they will all be rejected after this entry

--- a/pkg/storage/replica_application.go
+++ b/pkg/storage/replica_application.go
@@ -111,7 +111,7 @@ func (r *Replica) handleCommittedEntriesRaftMuLocked(
 			// Decode zero or more trivial entries into b.
 			var numEmptyEntries int
 			haveNonTrivialEntry, numEmptyEntries, toProcess, errExpl, err =
-				b.decode(ctx, toProcess, &b.decodeBuf)
+				b.decode(ctx, toProcess, &b.decodeBuf, r.store.TestingKnobs().MaxApplicationBatchSize)
 			if err != nil {
 				return stats, errExpl, err
 			}

--- a/pkg/storage/replica_application.go
+++ b/pkg/storage/replica_application.go
@@ -873,6 +873,11 @@ func (r *Replica) applyCmdAppBatch(
 	}
 	for ok := it.init(&b.cmdBuf); ok; ok = it.next() {
 		cmd := it.cur()
+		// Reset the context for already applied commands to ensure that
+		// reproposals at the same MaxLeaseIndex do not record into closed spans.
+		if cmd.proposedLocally() && cmd.proposal.applied {
+			cmd.ctx = ctx
+		}
 		for _, sc := range cmd.replicatedResult().SuggestedCompactions {
 			r.store.compactor.Suggest(cmd.ctx, sc)
 		}

--- a/pkg/storage/replica_application_result.go
+++ b/pkg/storage/replica_application_result.go
@@ -457,6 +457,9 @@ func (r *Replica) finishRaftCommand(ctx context.Context, cmd *cmdAppCtx) {
 	}
 
 	if cmd.proposedLocally() {
+		if cmd.raftCmd.MaxLeaseIndex != cmd.proposal.command.MaxLeaseIndex {
+			log.Fatalf(ctx, "finishing proposal with outstanding reproposal at a higher max lease index")
+		}
 		cmd.proposal.finishApplication(cmd.response)
 	} else if cmd.response.Err != nil {
 		log.VEventf(ctx, 1, "applying raft command resulted in error: %s", cmd.response.Err)

--- a/pkg/storage/replica_raft.go
+++ b/pkg/storage/replica_raft.go
@@ -435,27 +435,6 @@ func (r *Replica) handleRaftReadyRaftMuLocked(
 	leaderID := r.mu.leaderID
 	lastLeaderID := leaderID
 
-	// We defer the check to Replica.updateProposalQuotaRaftMuLocked because we need
-	// to check it in both cases, if hasReady is false and otherwise.
-	// If hasReady == false:
-	//     Consider the case when our quota is of size 1 and two out of three
-	//     replicas have committed one log entry while the third is lagging
-	//     behind. When the third replica finally does catch up and sends
-	//     along a MsgAppResp, since the entry is already committed on the
-	//     leader replica, no Ready is emitted. But given that the third
-	//     replica has caught up, we can release
-	//     some quota back to the pool.
-	// Otherwise:
-	//     Consider the case where there are two replicas and we have a quota
-	//     of size 1. We acquire the quota when the write gets proposed on the
-	//     leader and expect it to be released when the follower commits it
-	//     locally. In order to do so we need to have the entry 'come out of
-	//     raft' and in the case of a two node raft group, this only happens if
-	//     hasReady == true.
-	//     If we don't release quota back at the end of
-	//     handleRaftReadyRaftMuLocked, the next write will get blocked.
-	defer r.updateProposalQuotaRaftMuLocked(ctx, lastLeaderID)
-
 	err := r.withRaftGroupLocked(true, func(raftGroup *raft.RawNode) (bool, error) {
 		if err := r.mu.proposalBuf.FlushLockedWithRaftGroup(raftGroup); err != nil {
 			return false, err
@@ -472,6 +451,15 @@ func (r *Replica) handleRaftReadyRaftMuLocked(
 	}
 
 	if !hasReady {
+		// We must update the proposal quota even if we don't have a ready.
+		// Consider the case when our quota is of size 1 and two out of three
+		// replicas have committed one log entry while the third is lagging
+		// behind. When the third replica finally does catch up and sends
+		// along a MsgAppResp, since the entry is already committed on the
+		// leader replica, no Ready is emitted. But given that the third
+		// replica has caught up, we can release
+		// some quota back to the pool.
+		r.updateProposalQuotaRaftMuLocked(ctx, lastLeaderID)
 		return stats, "", nil
 	}
 
@@ -767,6 +755,18 @@ func (r *Replica) handleRaftReadyRaftMuLocked(
 	}); err != nil {
 		return stats, expl, errors.Wrap(err, expl)
 	}
+
+	// NB: All early returns other than the one due to not having a ready
+	// which also makes the below call are due to fatal errors.
+	// We must also update the proposal quota when have a ready; consider the
+	// case where there are two replicas and we have a quota of size 1. We
+	// acquire the quota when the write gets proposed on the leader and expect it
+	// to be released when the follower commits it locally. In order to do so we
+	// need to have the entry 'come out of raft' and in the case of a two node
+	// raft group, this only happens if hasReady == true. If we don't release
+	// quota back at the end of handleRaftReadyRaftMuLocked, the next write will
+	// get blocked.
+	r.updateProposalQuotaRaftMuLocked(ctx, lastLeaderID)
 	return stats, "", nil
 }
 

--- a/pkg/storage/replica_raft.go
+++ b/pkg/storage/replica_raft.go
@@ -258,8 +258,20 @@ func (r *Replica) evalAndPropose(
 //
 // The method hands ownership of the command over to the Raft machinery. After
 // the method returns, all access to the command must be performed while holding
-// Replica.mu and Replica.raftMu.
-func (r *Replica) propose(ctx context.Context, p *ProposalData) (int64, *roachpb.Error) {
+// Replica.mu and Replica.raftMu. If a non-nil error is returned the
+// MaxLeaseIndex is not updated.
+func (r *Replica) propose(ctx context.Context, p *ProposalData) (index int64, pErr *roachpb.Error) {
+
+	// If an error occurs reset the command's MaxLeaseIndex to its initial value.
+	// Failure to propose will propagate to the client. An invariant of this
+	// package is that proposals which are finished carry a raft command with a
+	// MaxLeaseIndex equal to the proposal command's max lease index.
+	defer func(prev uint64) {
+		if pErr != nil {
+			p.command.MaxLeaseIndex = prev
+		}
+	}(p.command.MaxLeaseIndex)
+
 	// Make sure the maximum lease index is unset. This field will be set in
 	// propBuf.Insert and its encoded bytes will be appended to the encoding
 	// buffer as a RaftCommandFooter.

--- a/pkg/storage/replica_test.go
+++ b/pkg/storage/replica_test.go
@@ -64,6 +64,7 @@ import (
 	"go.etcd.io/etcd/raft"
 	"go.etcd.io/etcd/raft/raftpb"
 	"go.etcd.io/etcd/raft/tracker"
+	"golang.org/x/net/trace"
 )
 
 // allSpans is a SpanSet that covers *everything* for use in tests that don't
@@ -11556,4 +11557,167 @@ func TestSplitSnapshotWarningStr(t *testing.T) {
 		"; r12/2 is being probed (may or may not need a Raft snapshot)",
 		splitSnapshotWarningStr(12, status),
 	)
+}
+
+// TestHighestMaxLeaseIndexReproposalFinishesCommand exercises a case where a
+// command is reproposed twice at different MaxLeaseIndex values to ultimately
+// fail with an error which cannot be reproposed (say due to a lease transfer
+// or change to the gc threshold). This test works to exercise the invariant
+// that when a proposal has been reproposed at different MaxLeaseIndex value
+// the client is ultimately acknowledged with an error from a reproposal with
+// the largest index. The test verfies this condition by asserting that the
+// span used to trace the execution of the proposal is not used after the
+// proposal has been finished.
+func TestHighestMaxLeaseIndexReproposalFinishesCommand(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	// Set the trace infrastructure to log if a span is used after being finished.
+	defer enableTraceDebugUseAfterFree()()
+
+	// Set logging up to a test specific directory.
+	scope := log.Scope(t)
+	defer scope.Close(t)
+
+	tc := testContext{}
+	ctx := context.Background()
+
+	stopper := stop.NewStopper()
+	defer stopper.Stop(ctx)
+	tc.manualClock = hlc.NewManualClock(123)
+	cfg := TestStoreConfig(hlc.NewClock(tc.manualClock.UnixNano, time.Nanosecond))
+	// Set up tracing.
+	tracer := tracing.NewTracer()
+	tracer.Configure(&cfg.Settings.SV)
+	cfg.AmbientCtx.Tracer = tracer
+
+	// Below we set txnID to the value of the transaction we're going to force to
+	// be proposed multiple times.
+	var txnID uuid.UUID
+	// In the TestingProposalFilter we populater cmdID with the id of the proposal
+	// which corresponds to txnID.
+	var cmdID storagebase.CmdIDKey
+	// After we evalAndPropose the command we populate prop with the ProposalData
+	// value to enable reproposing the same command more than once.
+	var prop *ProposalData
+	// seen is used to detect the first application of our proposal.
+	var seen bool
+	cfg.TestingKnobs = StoreTestingKnobs{
+		MaxApplicationBatchSize: 1,
+		// Set the TestingProposalFilter in order to know the CmdIDKey for our
+		// request by detecting its txnID.
+		TestingProposalFilter: func(args storagebase.ProposalFilterArgs) *roachpb.Error {
+			if args.Req.Header.Txn != nil && args.Req.Header.Txn.ID == txnID {
+				cmdID = args.CmdID
+			}
+			return nil
+		},
+		// Detect the application of the proposal to repropose it and also
+		// invalidate the lease.
+		TestingApplyFilter: func(args storagebase.ApplyFilterArgs) (retry int, pErr *roachpb.Error) {
+			if seen || args.CmdID != cmdID {
+				return 0, nil
+			}
+			seen = true
+			// Repropose on a separate location to not mess with the
+			// goldenProtosBelowRaft checks.
+			reproposed := make(chan struct{})
+			go func() {
+				if _, pErr := tc.repl.propose(prop.ctx, prop); pErr != nil {
+					panic(pErr)
+				}
+				close(reproposed)
+			}()
+			<-reproposed
+			tc.repl.mu.Lock()
+			defer tc.repl.mu.Unlock()
+			// Flush the proposalBuf to ensure that the reproposal makes it into the
+			// Replica's proposal map.
+			if err := tc.repl.mu.proposalBuf.flushLocked(); err != nil {
+				panic(err)
+			}
+			// Increase the lease sequence so that future reproposals will fail with
+			// NotLeaseHolderError. This mimics the outcome of a leaseholder change
+			// slipping in between the application of the first proposal and the
+			// reproposals.
+			tc.repl.mu.state.Lease.Sequence++
+			// This return value will force another retry which will carry a yet
+			// higher MaxLeaseIndex and will trigger our invariant violation.
+			return int(proposalIllegalLeaseIndex),
+				roachpb.NewErrorf("forced error that can be reproposed at a higher index")
+		},
+	}
+	tc.StartWithStoreConfig(t, stopper, cfg)
+	key := roachpb.Key("a")
+	lease, _ := tc.repl.GetLease()
+	txn := newTransaction("test", key, roachpb.NormalUserPriority, tc.Clock())
+	txnID = txn.ID
+	ba := roachpb.BatchRequest{
+		Header: roachpb.Header{
+			RangeID: tc.repl.RangeID,
+			Txn:     txn,
+		},
+	}
+	ba.Timestamp = txn.OrigTimestamp
+	ba.Add(&roachpb.PutRequest{
+		RequestHeader: roachpb.RequestHeader{
+			Key: key,
+		},
+		Value: roachpb.MakeValueFromBytes([]byte("val")),
+	})
+
+	// Hold the RaftLock to ensure that after evalAndPropose our proposal is in
+	// the proposal map. Entries are only removed from that map underneath raft.
+	// We want to grab the proposal so that we can shove in an extra reproposal
+	// while the first proposal is being applied.
+	tc.repl.RaftLock()
+	tracedCtx, cleanup := tracing.EnsureContext(ctx, cfg.AmbientCtx.Tracer, "replica send")
+	ch, _, _, pErr := tc.repl.evalAndPropose(tracedCtx, lease, &ba, &allSpans, endCmds{})
+	if pErr != nil {
+		t.Fatal(pErr)
+	}
+	errCh := make(chan *roachpb.Error)
+	go func() {
+		res := <-ch
+		cleanup()
+		errCh <- res.Err
+	}()
+
+	func() {
+		tc.repl.mu.Lock()
+		defer tc.repl.mu.Unlock()
+		if err := tc.repl.mu.proposalBuf.flushLocked(); err != nil {
+			t.Fatal(err)
+		}
+		tc.repl.refreshProposalsLocked(0, reasonNewLeaderOrConfigChange)
+		prop = tc.repl.mu.proposals[cmdID]
+	}()
+	tc.repl.RaftUnlock()
+
+	if pErr = <-errCh; !testutils.IsPError(pErr, "NotLeaseHolder") {
+		t.Fatal(pErr)
+	}
+
+	// Round trip another proposal through the replica to ensure that previously
+	// committed entries have been applied.
+	_, pErr = tc.repl.sendWithRangeID(ctx, tc.repl.RangeID, ba)
+	if pErr != nil {
+		t.Fatal(pErr)
+	}
+	log.Flush()
+
+	stopper.Quiesce(ctx)
+	entries, err := log.FetchEntriesFromFiles(0, math.MaxInt64, 1,
+		regexp.MustCompile("net/trace"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(entries) > 0 {
+		t.Fatalf("reused span after free: %v", entries)
+	}
+}
+
+func enableTraceDebugUseAfterFree() (restore func()) {
+	prev := trace.DebugUseAfterFinish
+	trace.DebugUseAfterFinish = true
+	return func() { trace.DebugUseAfterFinish = prev }
 }

--- a/pkg/storage/replica_test.go
+++ b/pkg/storage/replica_test.go
@@ -58,6 +58,7 @@ import (
 	"github.com/cockroachdb/logtags"
 	"github.com/gogo/protobuf/proto"
 	"github.com/kr/pretty"
+	opentracing "github.com/opentracing/opentracing-go"
 	"github.com/pkg/errors"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
@@ -11706,6 +11707,108 @@ func TestHighestMaxLeaseIndexReproposalFinishesCommand(t *testing.T) {
 	log.Flush()
 
 	stopper.Quiesce(ctx)
+	entries, err := log.FetchEntriesFromFiles(0, math.MaxInt64, 1,
+		regexp.MustCompile("net/trace"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(entries) > 0 {
+		t.Fatalf("reused span after free: %v", entries)
+	}
+}
+
+// TestLaterReproposalsDoNotReuseContext ensures that when commands are
+// reproposed more than once at the same MaxLeaseIndex and the first command
+// applies that the later reproposals do not log into the proposal's context
+// as its underlying trace span may already be finished.
+func TestLaterReproposalsDoNotReuseContext(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	// Set the trace infrastructure to log if a span is used after being finished.
+	defer enableTraceDebugUseAfterFree()()
+
+	tc := testContext{}
+	ctx := context.Background()
+
+	// Set logging up to a test specific directory.
+	scope := log.Scope(t)
+	defer scope.Close(t)
+
+	stopper := stop.NewStopper()
+	defer stopper.Stop(ctx)
+	cfg := TestStoreConfig(hlc.NewClock(hlc.UnixNano, time.Nanosecond))
+	// Set up tracing.
+	tracer := tracing.NewTracer()
+	tracer.Configure(&cfg.Settings.SV)
+	tracer.AlwaysTrace()
+	cfg.AmbientCtx.Tracer = tracer
+	tc.StartWithStoreConfig(t, stopper, cfg)
+	key := roachpb.Key("a")
+	lease, _ := tc.repl.GetLease()
+	txn := newTransaction("test", key, roachpb.NormalUserPriority, tc.Clock())
+	ba := roachpb.BatchRequest{
+		Header: roachpb.Header{
+			RangeID: tc.repl.RangeID,
+			Txn:     txn,
+		},
+	}
+	ba.Timestamp = txn.OrigTimestamp
+	ba.Add(&roachpb.PutRequest{
+		RequestHeader: roachpb.RequestHeader{
+			Key: key,
+		},
+		Value: roachpb.MakeValueFromBytes([]byte("val")),
+	})
+
+	// Hold the RaftLock to encourage the reproposals to occur in the same batch.
+	tc.repl.RaftLock()
+	sp := tracer.StartRootSpan("replica send", logtags.FromContext(ctx), tracing.RecordableSpan)
+	tracedCtx := opentracing.ContextWithSpan(ctx, sp)
+	// Go out of our way to enable recording so that expensive logging is enabled
+	// for this context.
+	tracing.StartRecording(sp, tracing.SingleNodeRecording)
+	ch, _, _, pErr := tc.repl.evalAndPropose(tracedCtx, lease, &ba, &allSpans, endCmds{})
+	if pErr != nil {
+		t.Fatal(pErr)
+	}
+	// Launch a goroutine to finish the span as soon as a result has been sent.
+	errCh := make(chan *roachpb.Error)
+	go func() {
+		res := <-ch
+		sp.Finish()
+		errCh <- res.Err
+	}()
+
+	// Flush the proposal and then repropose it twice.
+	// This test verifies that these later reproposals don't record into the
+	// tracedCtx after its span has been finished.
+	func() {
+		tc.repl.mu.Lock()
+		defer tc.repl.mu.Unlock()
+		if err := tc.repl.mu.proposalBuf.flushLocked(); err != nil {
+			t.Fatal(err)
+		}
+		tc.repl.refreshProposalsLocked(0, reasonNewLeaderOrConfigChange)
+		if err := tc.repl.mu.proposalBuf.flushLocked(); err != nil {
+			t.Fatal(err)
+		}
+		tc.repl.refreshProposalsLocked(0, reasonNewLeaderOrConfigChange)
+	}()
+	tc.repl.RaftUnlock()
+
+	if pErr = <-errCh; pErr != nil {
+		t.Fatal(pErr)
+	}
+	// Round trip another proposal through the replica to ensure that previously
+	// committed entries have been applied.
+	_, pErr = tc.repl.sendWithRangeID(ctx, tc.repl.RangeID, ba)
+	if pErr != nil {
+		t.Fatal(pErr)
+	}
+
+	stopper.Quiesce(ctx)
+	// Check and see if the trace package logged an error.
+	log.Flush()
 	entries, err := log.FetchEntriesFromFiles(0, math.MaxInt64, 1,
 		regexp.MustCompile("net/trace"))
 	if err != nil {

--- a/pkg/storage/testing_knobs.go
+++ b/pkg/storage/testing_knobs.go
@@ -195,6 +195,11 @@ type StoreTestingKnobs struct {
 	// This ensures the `*Replica` will be materialized on the Store when it
 	// returns.
 	ReplicaAddStopAfterLearnerSnapshot func() bool
+
+	// MaxApplicationBatchSize enforces a maximum size on application batches.
+	// This can be useful for testing conditions which require commands to be
+	// applied in separate batches.
+	MaxApplicationBatchSize int
 }
 
 // ModuleTestingKnobs is part of the base.ModuleTestingKnobs interface.

